nexus/src/db/queries/external_ip.rs

```diff
@@ -966,7 +1002,161 @@ mod tests {
         assert_eq!(ip.ip.ip(), first_range.first_address());
         assert_eq!(ip.first_port.0, 0);
         assert_eq!(ip.last_port.0, u16::MAX);
-        assert_eq!(ip.project_id, project_id);
+        assert_eq!(ip.project_id.unwrap(), project_id);
+
+        context.success().await;
+    }
+
+    #[tokio::test]
+    async fn test_next_external_ip_for_service() {
+        let context =
+            TestContext::new("test_next_external_ip_for_service").await;
+
+        // Create an IP pool without an associated project.
+        let rack_id = Uuid::parse_str(RACK_UUID).unwrap();
+        let ip_range = IpRange::try_from((
+            Ipv4Addr::new(10, 0, 0, 1),
+            Ipv4Addr::new(10, 0, 0, 2),
+        ))
+        .unwrap();
+        context.create_rack_ip_pool("p0", ip_range, rack_id).await;
+
+        // Allocate an IP address as we would for an external, rack-associated
+        // service.
+        let id1 = Uuid::new_v4();
+        let ip1 = context
+            .db_datastore
+            .allocate_service_ip(&context.opctx, id1, rack_id)
+            .await
+            .expect("Failed to allocate service IP address");
+        assert_eq!(ip1.kind, IpKind::Service);
+        assert_eq!(ip1.ip.ip(), IpAddr::V4(Ipv4Addr::new(10, 0, 0, 1)));
+        assert_eq!(ip1.first_port.0, 0);
+        assert_eq!(ip1.last_port.0, u16::MAX);
+        assert!(ip1.instance_id.is_none());
+        assert!(ip1.project_id.is_none());
+
+        // Allocate the next (last) IP address
+        let id2 = Uuid::new_v4();
+        let ip2 = context
+            .db_datastore
+            .allocate_service_ip(&context.opctx, id2, rack_id)
+            .await
+            .expect("Failed to allocate service IP address");
+        assert_eq!(ip2.kind, IpKind::Service);
+        assert_eq!(ip2.ip.ip(), IpAddr::V4(Ipv4Addr::new(10, 0, 0, 2)));
+        assert_eq!(ip2.first_port.0, 0);
+        assert_eq!(ip2.last_port.0, u16::MAX);
+        assert!(ip2.instance_id.is_none());
+        assert!(ip2.project_id.is_none());
+
+        // Once we're out of IP addresses, test that we see the right error.
+        let id3 = Uuid::new_v4();
+        let err = context
+            .db_datastore
+            .allocate_service_ip(&context.opctx, id3, rack_id)
+            .await
+            .expect_err("Should have failed to allocate after pool exhausted");
+        assert_eq!(
+            err,
+            Error::InvalidRequest {
+                message: String::from(
+                    // TODO: The error is a bit misleading; this isn't an IP
```

Comment:
Fair, it could just say "No External IP addresses available".

Reply:
Actually, thinking out loud, will users ever see this error message in the case that we fail to get an address for an Oxide service? For the most part, they won't be deciding to create new services like Nexus, the control plane itself should do that. One potential exception is during setup and bootstrapping, where we may need to alert the operator that we couldn't bring up enough Nexus instances (or any service) because there weren't enough IPs. Not sure.

Reply:
I think just omitting the "for new instance" leaves the mechanics accurate. If a user sees an unwrapped error that looks like:

```
- Failed to allocate instance
- No external IP addresses available
```

That's perfectly understandable, IMO. I think it's risky to make error messages dependent on the caller's context, which we don't know ahead of time.

Reply:
BTW, this error message is updated in https://github.com/oxidecomputer/omicron/pull/1635